### PR TITLE
[JN-946] enrollee invitation proof-of-concept

### PR DESCRIPTION
**Pull request by devonbush**
#### DESCRIPTION

This enables invitation emails to be sent to participants who have data in Juniper (such as from a migration), but do not have b2c accounts, allowing them to create their linked b2c account

image

To enable easy testing of this feature, I needed a way to repeatedly create an enrollee who did not have a b2c account.  So I introduced the concept of "keyed" populate users.  This lets us be sure that every time we populate demo, the `HDINVI` enrollee will be created with a new username and so will be able to be invited without conflicting with an existing b2c account.  The email generated looks like [your username]+[key]-[randomchars]@broadinstitute.org.

FUTURE WORK:
 1. b2c configuration to prefill the email address field (and maybe make it read-only)
 2. not hardcoding all "AD_HOC" emails to prompt for subject/body text
 3. some sort of indication in the data model (other than lastLogin) that a participantUser is (a) from a migration and (b) does not yet have a b2c account

#### TO TEST:  *(simple manual steps for confirming core behavior -- used for pre-release checks)*

1. redeploy ApiAdminApp
2. repopulate demo
3. make sure the ui-participant is running with b2c auth *on*
4. go to sandbox participant list `https://localhost:3000/demo/studies/heartdemo/env/sandbox/participants`
5. check-select the HDINVI enrollee, and press "Send email"
6. In the "email template" dropdown, select "Invitation email"
7. ignore the text fields, and press "send to 1 participant"
8. Confirm you get an email that looks like the screenshot above
9. click the link, confirm you are taken (after a redirect) to the "Verify your email" screen
10. (you won't be able to actually verify the email since demo b2c isn't yet configured)

### CODE REVIEW

**Review by connorlbark — APPROVED** [review]
Worked well for me - I assume it's expected that if you go through the full invitation process it doesn't work (I got a 500 error on every page since there wasn't a portal participant user). 👍  pending the failing test being fixed.

**Comment by devonbush** [comment]
hmmm... @connorlbark  where are you seeing there's no portal participant user?  when I do the populate demo, a portal participant user gets created for the 'invited' participant.  I wonder if there's some condition that can cause that populate to be flaky?

File: ui-participant/src/landing/registration/InvitationPage.tsx [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+      // we don't need to create a b2c account, just log the user in
+      const result = await Api.unauthedLogin(accountName)
+      loginUser(result, result.user.token)
+      navigate('/hub')
+      return
+    }
+    auth.signinRedirect({
+      redirectMethod: 'replace',
+      extraQueryParams: {
+        option: 'signup2',
```
Comment (MatthewBemis) [inline ui-participant/src/landing/registration/InvitationPage.tsx]: Is `signup2` a typo?

**Review by MatthewBemis — APPROVED** [review]
Looks good. One minor question

And as another data point, the portal participant user was successfully created when I tested this out

File: ui-participant/src/landing/registration/InvitationPage.tsx (hunk 1 above)
Comment (devonbush) [reply ui-participant/src/landing/registration/InvitationPage.tsx]: leftover from testing, good catch!

**Comment by devonbush** [comment]
security hotspot is using random number generator in populate file.  psuedo random is fine there since it's dev-only

**Pull request merged**